src/frontend/json.rs

```diff
@@ -0,0 +1,856 @@
+//! [`SeccompCmpOp`](enum.SeccompCmpOp.html),
+//! [`SeccompCmpArgLen`](enum.SeccompCmpArgLen.html).
+
+use std::collections::{BTreeMap, HashMap};
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::io::Read;
+use std::result;
+
+use crate::backend::{
+    Error as BackendError, SeccompAction, SeccompCmpArgLen, SeccompCmpOp, SeccompCondition,
+    SeccompFilter, SeccompRule, TargetArch,
+};
+use crate::syscall_table::SyscallTable;
+use serde::de::{self, Error as _, MapAccess, Visitor};
+use serde::{Deserialize, Deserializer};
+
+type Result<T> = result::Result<T, Error>;
+
+/// Error compiling JSON into IR.
+#[derive(Debug)]
+pub enum Error {
+    /// Backend error creating the `SeccompFilter` IR.
+    Backend(BackendError),
+    /// Error deserializing JSON.
+    SerdeJson(serde_json::Error),
+    /// Invalid syscall name for the given arch.
+    SyscallName(String, TargetArch),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            Backend(error) => write!(f, "{}", error),
+            SerdeJson(error) => {
+                write!(f, "Error parsing Json: {}", error)
+            }
+            SyscallName(syscall_name, arch) => write!(
+                f,
+                "Invalid syscall name: {} for given arch: {:?}.",
+                syscall_name, arch
+            ),
+        }
+    }
+}
+
+/// Deserializable object that represents the Json filter file.
+// Need the 'newtype' pattern so that we can implement a custom deserializer.
+pub(crate) struct JsonFile(pub HashMap<String, JsonFilter>);
+
+// Implement a custom deserializer, that returns an error for duplicate thread keys.
+impl<'de> Deserialize<'de> for JsonFile {
+    fn deserialize<D>(deserializer: D) -> result::Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        struct JsonFileVisitor;
+
+        impl<'d> Visitor<'d> for JsonFileVisitor {
+            type Value = HashMap<String, JsonFilter>;
+
+            fn expecting(&self, f: &mut fmt::Formatter<'_>) -> result::Result<(), fmt::Error> {
+                f.write_str("a map of filters")
+            }
+
+            fn visit_map<M>(self, mut access: M) -> result::Result<Self::Value, M::Error>
+            where
+                M: MapAccess<'d>,
+            {
+                let mut values = Self::Value::with_capacity(access.size_hint().unwrap_or(0));
+
+                while let Some((key, value)) = access.next_entry()? {
+                    if values.insert(key, value).is_some() {
+                        return Err(M::Error::custom("duplicate filter key"));
+                    };
+                }
+
+                Ok(values)
+            }
+        }
+        Ok(JsonFile(deserializer.deserialize_map(JsonFileVisitor)?))
+    }
+}
+
+/// Dummy placeholder type for a JSON comment. Holds no value.
+#[derive(PartialEq, Debug, Clone)]
```

Comment:
Can you add some more details in comments about why this define in needed and how it helps?